cpp/src/parquet/arrow/schema.cc

```diff
@@ -984,21 +992,51 @@ Result<bool> ApplyOriginalMetadata(const Field& origin_field, SchemaField* infer
   bool modified = false;
 
   auto& origin_type = origin_field.type();
+  auto inferred_type = inferred->field->type();
 
   if (origin_type->id() == ::arrow::Type::EXTENSION) {
     const auto& ex_type = checked_cast<const ::arrow::ExtensionType&>(*origin_type);
-    auto origin_storage_field = origin_field.WithType(ex_type.storage_type());
+    if (inferred_type->id() != ::arrow::Type::EXTENSION &&
+        ex_type.extension_name() == ::arrow::extension::JsonExtensionType::type_name()) {
+      // Schema mismatch.
+      //
+      // Arrow extensions are DISABLED in Parquet.
+      // origin_type is ::arrow::extension::json()
+      // inferred_type is ::arrow::binary()
+      //
+      // Origin type is restored as Arrow should be considered the source of truth.
+      DCHECK_EQ(inferred_type->id(), ::arrow::Type::BINARY);
+      inferred->field = inferred->field->WithType(origin_type);
+      RETURN_NOT_OK(ApplyOriginalStorageMetadata(origin_field, inferred));
+    } else {
+      auto origin_storage_field = origin_field.WithType(ex_type.storage_type());
 
-    // Apply metadata recursively to storage type
-    RETURN_NOT_OK(ApplyOriginalStorageMetadata(*origin_storage_field, inferred));
+      // Apply metadata recursively to storage type
+      RETURN_NOT_OK(ApplyOriginalStorageMetadata(*origin_storage_field, inferred));
 
-    // Restore extension type, if the storage type is the same as inferred
-    // from the Parquet type
-    if (ex_type.storage_type()->Equals(*inferred->field->type())) {
-      inferred->field = inferred->field->WithType(origin_type);
+      // Restore extension type, if the storage type is the same as inferred
+      // from the Parquet type
+      if (ex_type.storage_type()->Equals(*inferred->field->type())) {
+        inferred->field = inferred->field->WithType(origin_type);
+      }
     }
     modified = true;
   } else {
+    if (inferred_type->id() == ::arrow::Type::EXTENSION) {
+      const auto& ex_type = checked_cast<const ::arrow::ExtensionType&>(*inferred_type);
+      if (ex_type.extension_name() ==
+          ::arrow::extension::JsonExtensionType::type_name()) {
+        // Schema mismatch.
+        //
+        // Arrow extensions are ENABLED in Parquet.
+        // origin_type is ::arrow::binary()
+        // inferred_type is ::arrow::extension::json()
+        //
+        // Origin type is restored as Arrow should be considered the source of truth.
```

Comment:
I'm unsure that it is really desirable, since this is also losing information about the logical type.

Reply:
Perhaps we can raise for now?

Reply:
I'm not sure what to do here @pitrou. What would the preferred behavior be?

Reply:
We can just ignore this case here IMHO.

Reply:
https://github.com/apache/arrow/blob/6c17b794509d3931225cf295ae864204162c786f/cpp/src/parquet/arrow/schema.cc#L428-L433

@rok  is right

I think we can support serialize whole schema here, and during read, if we don't recognize that, we can use underlying storage? ( this might be in separate patch )

Reply:
This would only happen if a Parquet writer wrote a Arrow utf8 array (_not_ JSON) as a JSON Parquet logical type. Why would it do that?

Reply:
I don't have a strong opinion on this either way. The example I can think of is a another implementation without canonical types writing this.

Reply:
> The example I can think of is a another implementation without canonical types writing this.

Then would we not want to have a JSON array, rather than plain UTF8?

Reply:
> Then would we not want to have a JSON array, rather than plain UTF8?

I suppose so! Changed:

```diff
-        inferred->field = inferred->field->WithType(origin_type);
+        inferred->field = inferred->field->WithType(inferred_type);
```

Reply:
Which is essentially a no-op, isn't it? (`inferred_type` is already `inferred->field`'s type)
So we can just remove this entire code path, it seems.